### PR TITLE
Implemented solution for hw24

### PR DESCRIPTION
**Pull request by andriylagun**

### CODE REVIEW

File: src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        return Storage.productStorage;
+    }
+
+    @Override
+    public Product update(Product product) {
+        return Storage.productStorage.stream()
+                .filter(p -> p.getId().equals(product.getId()))
+                .map(p -> p = product)
+                .findFirst()
+                .orElseThrow();
```
Comment (Rommelua) [inline src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java]: In this method you doesn't update product in storage.

File: src/main/java/Application.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        productService.create(new Product("Xiaomi Redmi Note 6", 400.0));
+        productService.create(new Product("Nokia 1100", 10.50));
+        System.out.println("All products:");
+        for (Product product : productService.getAll()) {
+            System.out.println(product);
+        }
+        System.out.println("Xiaomi Redmi Note 6 update:");
+        Product greenTeaProduct = productService.get(XIAOMI_ID);
+        greenTeaProduct.setPrice(350.00);
+        productService.update(greenTeaProduct);
```
Comment (Rommelua) [inline src/main/java/Application.java]: You have changed price when invoked setPrice(), not update() 
Use to find mistake
```
Product productToUpdate = new Product("Xiaomi Redmi Note 6", 400.0);
        productToUpdate.setId(2L);
        productService.update(productToUpdate);
```

File: src/main/java/com/internet/shop/dao/impl/UserDaoImpl.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+@Dao
+public class UserDaoImpl implements UserDao {
+
+    @Override
+    public boolean addUser(User user) {
+        return Storage.userStorage.add(user);
+    }
+
+    @Override
+    public User getUserById(long id) {
```
Comment (Rommelua) [inline src/main/java/com/internet/shop/dao/impl/UserDaoImpl.java]: We should return Optional here

File: src/main/java/com/internet/shop/model/Order.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
+package com.internet.shop.model;
+
+import java.util.List;
+import lombok.Data;
+
+@Data
+public class Order {
+    private static long idCount = 1;
```
Comment (Rommelua) [inline src/main/java/com/internet/shop/model/Order.java]: Of course, this logic will work to, but in task we were asked to make increment in Storage...

**Comment by Rommelua** [comment]
Good job!

File: .travis.yml [hunk 5]
```diff
@@ -0,0 +1,8 @@
+language: java
+jdk:
+  - openjdk11
+script:
+  - mvn clean install
+cache:
+  directories:
+    - $HOME/.m2
```
Comment (KseniiaMakarova) [inline .travis.yml]: your master already contains this file

File: pom.xml [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    </properties>
+
+    <dependencies>
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <version>4.12</version>
+            <scope>test</scope>
+        </dependency>
+        <!-- https://mvnrepository.com/artifact/org.projectlombok/lombok -->
```
Comment (KseniiaMakarova) [inline pom.xml]: let's remove junit dependency and commented line

File: src/main/java/Application.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
+import com.internet.shop.lib.Injector;
+import com.internet.shop.model.Product;
+import com.internet.shop.service.ProductService;
+
+public class Application {
```
Comment (KseniiaMakarova) [inline src/main/java/Application.java]: put this class inside `com.internet.shop`

File: src/main/java/com/internet/shop/dao/UserDao.java [hunk 8]
```diff
@@ -0,0 +1,17 @@
+package com.internet.shop.dao;
+
+import com.internet.shop.model.User;
+import java.util.List;
+import java.util.Optional;
+
+public interface UserDao {
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/dao/UserDao.java:3]: currently we don't need to write methods here
please remove this for now

File: src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+        return Storage.productStorage;
+    }
+
+    @Override
+    public Product update(Product product) {
+        Product updatedProduct = get(product.getId()).get();
+        updatedProduct.setName(product.getName());
+        updatedProduct.setPrice(product.getPrice());
+        return updatedProduct;
+    }
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java]: the logic of this method is to update all possible matches, not only the first one

let's create IntStream of indices, than filter out only those indices on which there are elements with the same id
then for each of the remaining indices, replace the element with our `product`

and let's return the new product (actually return the same object that we receive)

File: src/main/java/com/internet/shop/dao/impl/UserDaoImpl.java [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+import com.internet.shop.dao.UserDao;
+import com.internet.shop.db.Storage;
+import com.internet.shop.lib.Dao;
+import com.internet.shop.model.User;
+import java.util.List;
+import java.util.Optional;
+
+@Dao
+public class UserDaoImpl implements UserDao {
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/dao/impl/UserDaoImpl.java:7]: let's remove this

File: src/main/java/com/internet/shop/db/Storage.java [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+
+public class Storage {
+    public static final List<User> userStorage = new ArrayList<>();
+    public static final List<Product> productStorage = new ArrayList<>();
+    public static final List<ShoppingCart> shoppingCartStorage = new ArrayList<>();
+    public static final List<Order> orderStorage = new ArrayList<>();
+    private static long userId = 1;
+    private static long productId = 1;
+    private static long shoppingCartId = 1;
+    private static long orderId = 1;
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/db/Storage.java:18]:
```suggestion
    private static Long userId = 0L;
    private static Long productId = 0L;
    private static Long shoppingCartId = 0L;
    private static Long orderId = 0L;
```

File: src/main/java/com/internet/shop/db/Storage.java [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+    public static void addUser(User user) {
+        user.setId(userId++);
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/db/Storage.java:21]:
```suggestion
        user.setId(++userId);
```
same for other methods

File: src/main/java/com/internet/shop/model/Order.java [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+import lombok.Data;
+
+@Data
+public class Order {
+    private Long id;
+    private List<Product> products;
+    private Long userId;
+
+    public Order(Long userId) {
+        this.products = products;
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/model/Order.java]: what does this line mean?

File: src/main/java/com/internet/shop/model/ShoppingCart.java [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+import lombok.Data;
+
+@Data
+public class ShoppingCart {
+    private Long id;
+    private List<Product> products;
+    private Long userId;
+
+    public ShoppingCart(Long userId) {
+        this.userId = userId;
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/model/ShoppingCart.java:15]: let's initialize `products` here as well

File: src/main/java/com/internet/shop/service/impl/ProductServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+    private ProductDao productDao;
+
+    @Override
+    public Product create(Product product) {
+        return productDao.create(product);
+    }
+
+    @Override
+    public Product get(Long id) {
+        return productDao.get(id).orElseThrow();
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/service/impl/ProductServiceImpl.java]: let's use `get()` for now

File: src/main/java/Application.java [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+        productService.create(new Product("Iphone", 800.50));
+        productService.create(new Product("Xiaomi Redmi Note 6", 400.0));
+        productService.create(new Product("Nokia 1100", 10.50));
+        System.out.println("All products:");
+        for (Product product : productService.getAll()) {
+            System.out.println(product);
+        }
+        System.out.println("Xiaomi Redmi Note 6 update:");
+        Product productToUpdate = new Product("Xiaomi Redmi Note 6", 350.0);
+        productToUpdate.setId(2L);
```
Comment (KseniiaMakarova) [inline src/main/java/Application.java]: let's save our products in variables
and then call `someProduct.getId()` if needed

File: src/main/java/Application.java [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+        productService.update(productToUpdate);
+        System.out.println(productService.get(XIAOMI_ID));
```
Comment (KseniiaMakarova) [inline src/main/java/Application.java]: same here

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request merged**